From 2c66600edb0c7f2196644f1754e23c5950d10d88 Mon Sep 17 00:00:00 2001
From: Thomas Colliers <mail@thomascolliers.com>
Date: Mon, 10 Aug 2026 22:37:17 +0200
Subject: [PATCH 1/2] arm64: dts: qcom: milos: Add CAMSS

Add the camss@ac13000 node with its register regions, interrupts, clocks,
interconnects, IOMMUs and power domain. Describe the four CSI input ports
and the default MCLK1 pin configuration.

Leave CAMSS disabled and the ports without endpoints. Boards provide the
CSIPHY supplies and sensor connections when enabling the subsystem. Keep
sensor-specific wiring out of the SoC description so it can be reused by
other Milos camera configurations.

Assisted-by: pi:claude-opus-5
Suggested-by: Jorijn van der Graaf <jorijnvdgraaf@catcrafts.net>
Signed-off-by: Thomas Colliers <mail@thomascolliers.com>
---
 arch/arm64/boot/dts/qcom/milos.dtsi | 116 ++++++++++++++++++++++++++++
 1 file changed, 116 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/milos.dtsi b/arch/arm64/boot/dts/qcom/milos.dtsi
index aad8bf678916..5f9e6b640db6 100644
--- a/arch/arm64/boot/dts/qcom/milos.dtsi
+++ b/arch/arm64/boot/dts/qcom/milos.dtsi
@@ -1954,6 +1954,115 @@ videocc: clock-controller@aaf0000 {
 			#power-domain-cells = <1>;
 		};
 
+		camss: camss@ac13000 {
+			compatible = "qcom,milos-camss";
+
+			reg = <0x0 0x0ac13000 0x0 0x1000>,
+			      <0x0 0x0ac62000 0x0 0x1000>,
+			      <0x0 0x0ac71000 0x0 0x1000>,
+			      <0x0 0x0ac80000 0x0 0x1000>,
+			      <0x0 0x0ace4000 0x0 0x2000>,
+			      <0x0 0x0ace6000 0x0 0x2000>,
+			      <0x0 0x0ace8000 0x0 0x2000>,
+			      <0x0 0x0acea000 0x0 0x2000>,
+			      <0x0 0x0ac63800 0x0 0xb800>,
+			      <0x0 0x0ac72800 0x0 0xb800>,
+			      <0x0 0x0ac81800 0x0 0xb800>;
+			reg-names = "top",
+				    "csid0", "csid1", "csid2",
+				    "csiphy0", "csiphy1", "csiphy2", "csiphy3",
+				    "vfe0", "vfe1", "vfe2";
+
+			clocks = <&camcc CAM_CC_CPAS_AHB_CLK>,
+				 <&camcc CAM_CC_SOC_AHB_CLK>,
+				 <&camcc CAM_CC_CSIPHY0_CLK>,
+				 <&camcc CAM_CC_CSI0PHYTIMER_CLK>,
+				 <&camcc CAM_CC_CSIPHY1_CLK>,
+				 <&camcc CAM_CC_CSI1PHYTIMER_CLK>,
+				 <&camcc CAM_CC_CSIPHY2_CLK>,
+				 <&camcc CAM_CC_CSI2PHYTIMER_CLK>,
+				 <&camcc CAM_CC_CSIPHY3_CLK>,
+				 <&camcc CAM_CC_CSI3PHYTIMER_CLK>,
+				 <&camcc CAM_CC_TFE_0_AHB_CLK>,
+				 <&camcc CAM_CC_TFE_1_AHB_CLK>,
+				 <&camcc CAM_CC_TFE_2_AHB_CLK>,
+				 <&camcc CAM_CC_TFE_0_CSID_CLK>,
+				 <&camcc CAM_CC_TFE_1_CSID_CLK>,
+				 <&camcc CAM_CC_TFE_2_CSID_CLK>,
+				 <&camcc CAM_CC_TFE_0_CPHY_RX_CLK>,
+				 <&camcc CAM_CC_TFE_1_CPHY_RX_CLK>,
+				 <&camcc CAM_CC_TFE_2_CPHY_RX_CLK>,
+				 <&camcc CAM_CC_TFE_0_CLK>,
+				 <&camcc CAM_CC_TFE_1_CLK>,
+				 <&camcc CAM_CC_TFE_2_CLK>,
+				 <&camcc CAM_CC_CAMNOC_RT_AXI_CLK>,
+				 <&camcc CAM_CC_CAMNOC_NRT_AXI_CLK>,
+				 <&camcc CAM_CC_CAMNOC_AXI_HF_CLK>,
+				 <&camcc CAM_CC_CAMNOC_AXI_SF_CLK>;
+			clock-names = "cpas_ahb",
+				      "soc_ahb",
+				      "csiphy0", "csiphy0_timer",
+				      "csiphy1", "csiphy1_timer",
+				      "csiphy2", "csiphy2_timer",
+				      "csiphy3", "csiphy3_timer",
+				      "tfe0_ahb", "tfe1_ahb", "tfe2_ahb",
+				      "csi0", "csi1", "csi2",
+				      "vfe0_cphy_rx", "vfe1_cphy_rx", "vfe2_cphy_rx",
+				      "vfe0", "vfe1", "vfe2",
+				      "camnoc_rt_axi", "camnoc_nrt_axi",
+				      "camnoc_axi_hf", "camnoc_axi_sf";
+
+			interrupts = <GIC_SPI 464 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 466 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 203 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 477 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 478 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 479 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 448 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 602 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 604 IRQ_TYPE_EDGE_RISING 0>,
+				     <GIC_SPI 688 IRQ_TYPE_EDGE_RISING 0>;
+			interrupt-names = "csid0", "csid1", "csid2",
+					  "csiphy0", "csiphy1", "csiphy2", "csiphy3",
+					  "vfe0", "vfe1", "vfe2";
+
+			interconnects = <&gem_noc MASTER_APPSS_PROC QCOM_ICC_TAG_ACTIVE_ONLY
+					 &cnoc_cfg SLAVE_CAMERA_CFG QCOM_ICC_TAG_ACTIVE_ONLY>,
+					<&mmss_noc MASTER_CAMNOC_HF QCOM_ICC_TAG_ALWAYS
+					 &mc_virt SLAVE_EBI1 QCOM_ICC_TAG_ALWAYS>,
+					<&mmss_noc MASTER_CAMNOC_SF QCOM_ICC_TAG_ALWAYS
+					 &mc_virt SLAVE_EBI1 QCOM_ICC_TAG_ALWAYS>;
+			interconnect-names = "ahb", "hf_mnoc", "sf_mnoc";
+
+			iommus = <&apps_smmu 0x0880 0x20>,
+				 <&apps_smmu 0x08a0 0x20>;
+
+			power-domains = <&camcc CAM_CC_CAMSS_TOP_GDSC>;
+
+			status = "disabled";
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				port@0 {
+					reg = <0>;
+				};
+
+				port@1 {
+					reg = <1>;
+				};
+
+				port@2 {
+					reg = <2>;
+				};
+
+				port@3 {
+					reg = <3>;
+				};
+			};
+		};
+
 		cci0: cci@ac15000 {
 			compatible = "qcom,milos-cci", "qcom,msm8996-cci";
 			reg = <0x0 0x0ac15000 0x0 0x1000>;
@@ -2474,6 +2583,13 @@ data-pins {
 				};
 			};
 
+			cam_mclk1_default: cam-mclk1-default-state {
+				pins = "gpio84";
+				function = "cam_mclk";
+				drive-strength = <4>;
+				bias-disable;
+			};
+
 			cci0_0_default: cci0-0-default-state {
 				sda-pins {
 					pins = "gpio88";
